### ISSUE TITLE
Mapping static video titles to playlist download

### ISSUE THREAD

**Issue by leonbeon** [yt-dlp/yt-dlp#9252]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm asking a question and **not** reporting a bug or requesting a feature
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Please make sure the question is worded well enough to be understood

I want to download a playlist of videos. Through some custom logic, I have determined the video file names beforehand. Is there any way to map them to the playlist items as they are downloaded? For a single video, this is simple enough, e.g.

(Using JS as an example)
```js
const name = "my_predetermined_name";
const outputTemplate = `${name}.%(ext)s`; // later using `-o ${outputTemplate}` as CLI argument
```

If I read the docs correctly, there is no way to generate a conditional template (`if(playlist_index==1){${name1}}.$(ext)s` or something horrible like that).

Is there any way to achieve this with command line arguments? I can imagine this to work with some kind of post-processing filter, e.g. naming all videos with `%(playlist_index)s.%(ext)s` to keep them unique and then mapping my file names after everything is downloaded.

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by pukkandan** [comment]
No, there is no direct way. It is possible to hack something together with `--parse-metadata` but that will be unnecessarily complex and unmaintainable

> e.g. naming all videos with `%(playlist_index)s.%(ext)s` to keep them unique and then mapping my file names after everything is downloaded.

This is the easiest solution imo

Another way is to get all the items with `--flat-playlist --print webpage_url` first and then download each separately. This won't have any significant overhead as opposed to `-I`

**Comment by leonbeon** [comment]
> Another way is to get all the items with `--flat-playlist --print webpage_url` first and then download each separately. This won't have any significant overhead as opposed to `-I`

I like this approach more, since I want to avoid downloading files and renaming them after the fact. However, `webpage_url` fails with playlist content that has the same URL (e.g. Instagram carousels). `--print urls` would work if the requested format is defined beforehand (via `-F`), right?

**Comment by pukkandan** [comment]
Can you give an example link?

**Comment by leonbeon** [comment]
Sure thing! This may not be the best example because there are pictures inbetween, but I guess they get ignored anyway: Take this instagram post with multiple videos (first and last slide): `https://www.instagram.com/p/C3k6_bnxVWu/`

```bash
> yt-dlp --flat-playlist --print webpage_url "https://www.instagram.com/p/C3k6_bnxVWu/"

WARNING: [Instagram] C3k6_bnxVWu: No csrf token set by Instagram API
https://www.instagram.com/p/C3k6_bnxVWu/
https://www.instagram.com/p/C3k6_bnxVWu/
```

while `--print urls` shows

```bash
> yt-dlp --flat-playlist --print urls "https://www.instagram.com/p/C3k6_bnxVWu/"

WARNING: [Instagram] C3k6_bnxVWu: No csrf token set by Instagram API
https://scontent-dus1-1.cdninstagram.com/v/t66.30100-16/169633756_1419297962336041_1499259416664075452_n.mp4?_nc_ht=scontent-dus1-1.cdninstagram.com&_nc_cat=101&_nc_ohc=iolwURmgLTEAX8ucamA&edm=AP_V10EBAAAA&ccb=7-5&oh=00_AfCvihJEKF0X-d-wH1Cvn-k6iKXs8C4LFvXooSUmODg21g&oe=65DE71DC&_nc_sid=2999b8
https://scontent-dus1-1.cdninstagram.com/v/t50.2886-16/417752257_7573037412735833_8734672007906710118_n.mp4?_nc_ht=scontent-dus1-1.cdninstagram.com&_nc_cat=105&_nc_ohc=kdop61KCa0kAX_Tvfm4&edm=AP_V10EBAAAA&ccb=7-5&oh=00_AfB1S841ERVk4zBr6lgq-QxGPotmJXKxsDSuHSsoQPUuwg&oe=65DE60B1&_nc_sid=2999b8
```

pointing to the video files directly. So for the purposes of differentiating between different videos in this "playlist", you can't really use `webpage_url`.

**Comment by pukkandan** [comment]
You are correct. `webpage_url` is supposed to return the video only as per docs. But the video has no URL and so core code copies the playlist URL into it, which is a bug.

In this case, should we just make webage_url empty?, or should core code set something like `webpage_url=smuggle_url(playlist_url, {'item': I})`? Then we can parse it back on re-attempt. opinions? @bashonly @Grub4K
